src/results/converters.rs

```diff
@@ -0,0 +1,45 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#[inline]
+pub fn hex_char_to_bin(c: char) -> &'static str {
+    match c {
```

Comment:
This is fine, but it might be a really fun opportunity to write a constant expression LUT generator function :)

Reply:
I leveraged [lazy_static](https://crates.io/crates/lazy_static) to generate a static lookup table in: https://github.com/Qiskit/qiskit-terra/pull/8051/commits/5c815108aba7d1ec3770440383bf45af455337df I need to benchmark it and look into expanding it to support larger chunks. But this might be a good enough start as it should eliminate most of the runtime branching.

Reply:
I tried playing with adding chunking in groups of 4 and it was slower than doing this. I think this is coming down to needing to do intermediate allocations in how I could get it to work. At this point I think just doing a single element lookup table is probably sufficient. If we end up hitting bottlenecks down the road I feel like we can revisit this easily enough as it's not a big deal to improve the internal implementation down the road.

 For benchmarking I compared prior to https://github.com/Qiskit/qiskit-terra/commit/5c815108aba7d1ec3770440383bf45af455337df with https://github.com/Qiskit/qiskit-terra/commit/5c815108aba7d1ec3770440383bf45af455337df and my local chunked implementation using:

```python
import time
import random
from qiskit.result import marginal_memory

random.seed(42)

memory = [hex(random.randint(0, 4096)) for _ in range(500000)]
start = time.perf_counter()
res = marginal_memory(memory, indices=[0, 3, 5, 9])
stop = time.perf_counter()
print(stop - start)
```

The geometric mean of each implementation over 10 trials each was:

```
match: 0.08678476060453334
LUT: 0.08359493472968436
Chunked LUT: 0.10288708564573844
```